agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -2313,6 +2319,66 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
     return send_cmdu_to_bus(cmdu_tx, src_mac, bridge_info.mac);
 }
 
+bool backhaul_manager::handle_ap_metrics_query(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                               const std::string &src_mac)
+{
+    std::vector<sMacAddr> bssid;
+    const auto mid            = cmdu_rx.getMessageId();
+    auto ap_metrics_query_tlv = cmdu_rx.getClass<wfa_map::tlvApMetricQuery>();
+    if (!ap_metrics_query_tlv) {
+        LOG(ERROR) << "AP Metrics Query CMDU mid=" << mid << " does not have AP Metric Query TLV";
+        return false;
+    }
+    for (int bssid_idx = 0; bssid_idx < ap_metrics_query_tlv->bssid_list_length(); bssid_idx++) {
+        auto bssid_tupl = ap_metrics_query_tlv->bssid_list(bssid_idx);
+        if (!std::get<0>(bssid_tupl)) {
+            LOG(ERROR) << "Failed to get bssid " << bssid_idx << " from AP_METRICS_QUERY";
+            return false;
+        }
+        bssid.push_back(std::get<1>(bssid_tupl));
+        LOG(DEBUG) << "Received AP_METRICS_QUERY_MESSAGE, mid=" << std::hex << int(mid)
+                   << "  bssid " << std::get<1>(bssid_tupl);
+    }
+
+    auto cmdu_header = cmdu_tx.create(0, ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE);
+
+    if (!cmdu_header) {
+        LOG(ERROR) << "Failed building IEEE1905 AP_METRICS_QUERY_MESSAGE";
+        return false;
+    }
+    auto ap_metrics_response_tlv = cmdu_tx.addClass<wfa_map::tlvApMetricQuery>();
+    ap_metrics_response_tlv      = cmdu_rx.getClass<wfa_map::tlvApMetricQuery>();
+    for (auto socket : slaves_sockets) {
+        for (auto mac : bssid) {
+            if (mac == socket->radio_mac)
+                LOG(DEBUG) << "Sending AP_METRICS_QUERY_MESSAGE message to son_slave, bssid: "
+                           << std::hex << network_utils::mac_to_string(mac);
+            message_com::send_cmdu(socket->slave, cmdu_tx);
+        }
+    }
+    return true;
+}
+
+bool backhaul_manager::handle_ap_metrics_response(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                                  const std::string &src_mac)
+{
+    const auto mid            = cmdu_rx.getMessageId();
+    auto ap_metrics_query_tlv = cmdu_rx.getClass<wfa_map::tlvApMetric>();
+
+    auto cmdu_header = cmdu_tx.create(mid, ieee1905_1::eMessageType::AP_METRICS_RESPONSE_MESSAGE);
+
+    if (!cmdu_header) {
+        LOG(ERROR) << "Failed building IEEE1905 AP_METRICS_RESPONSE_MESSAGE";
+        return false;
+    }
+
+    auto ap_metrics_response_tlv = cmdu_tx.addClass<wfa_map::tlvApMetric>();
+    ap_metrics_response_tlv      = ap_metrics_query_tlv;
+    LOG(DEBUG) << "Sending AP_METRICS_RESPONSE_MESSAGE, mid=" << std::hex << int(mid);
+
+    return send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
```

Comment:
I think that backhaul manager should wait until responses from all the slaves had been received. At that moment, backhaul manager should create and send a new response message with metrics received from all requested radios.

Reply:
Implemented this mechanism in the last update

Reply:
> When we receive a response that is not expected, it is forwarded as-is to the controller.

I'd ignore this one

Reply:
We also have to consider the case of periodically reporting AP metrics (without a query, no mid)